### ISSUE TITLE
provide a way for test collector to find custom test suites

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7250]
|[<img alt="keturn's avatar" src="https://avatars.githubusercontent.com/u/83819?s=50" width="50" height="50">](https://github.com/keturn)| @keturn reported|
|-|-|
|Trac ID|trac#1638|
|Type|enhancement|
|Created|2006-04-13 00:43:18Z|

We have some data-driven tests here, which we turn in to unittest-compatible tests by writing a bit of code to stuff them in to TestSuites.  It'd be nice if there were a way to expose these custom TestSuites to trial's test collector.

Last time I chatted with people about this, there wasn't clearly an existing convention for that.  It probably ends up looking something like "if a test module has a function with the magic name createTests(), trial runs it and collects the returned tests or suites."

(You can probably convince me to provide a patch.  Let me know if there are behaviors that are particularly desirable or unacceptable.)

<details><summary>Searchable metadata</summary>

```
trac-id__1638 1638
type__enhancement enhancement
reporter__acapnotic acapnotic
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1144888998000000 1144888998000000
changetime__1155508580000000 1155508580000000
version__ 
owner__ 
cc__spiv
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

I recommend "test_suite" rather than "createTests", as that's an existing convention used by e.g. the Zope 3 test runner.  I agree that Trial should support custom suites one way or another.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to spiv|
|-|-|

I think this is done in [branches/suite-1638.](branches/suite-1638.)
spiv, wanna revie?

**Comment by twisted-trac** [comment]
|[<img alt="keturn's avatar" src="https://avatars.githubusercontent.com/u/83819?s=50" width="50" height="50">](https://github.com/keturn)|@keturn commented|
|-|-|

I have a module with both standard TestCases and custom TestSuites,
the r16711 will only call test_suite, not pick up the standard TestCases.
well, that's okay, as long as I can use a test loader in my definition of test_suite...

but calling loadModule in r16711 would just result in calling test_suite again which would call loadModule until recursion limit reached.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to spiv|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

Review tickets should be "highest" priority so they don't languish.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @radix|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

I can see this being pretty useful, and the change is fine as far as changes go, but I'm really not sure about breaking the coding standard with "test_suite". Is this a Pyunit compatibility thing?

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

Zope 3 and bzr's test runners both look for "test_suite" functions in modules.  I'm not sure if there's a deeper reason for that or not, but it does seem to be at least a bit of a precedent.

If we do something else in Trial, I'd still like there to be an easy way to make Trial look for "test_suite" functions, so I can use it with Zope 3 and bzr tests...

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub commented|
|-|-|

This isn't documented anywhere. I would prefer this wasn't merged until this is documented. Please add docstrings to the tests, and please add a docstring to loadModule.

I tried to find the docs for writing tests, but the only thing I found in svn was trunk/doc/core/howto/policy/test-standard.xhtml and that document isn't appropriate for documenting this change I don't feel. Perhaps a new ticket would be appropriate for the writing of a new document detailing the different ways it's possible to write unit tests.

I like how simple the change is. It's really quite neat and simple.

Be warned, merging will cause a conflict because other functions were added to the end of the file. you'll have to remove the <<<, === and >>> lines before commiting. Please make sure to run trial twisted.trial to make sure before committing.

**Comment by twisted-trac** [comment]
|[<img alt="keturn's avatar" src="https://avatars.githubusercontent.com/u/83819?s=50" width="50" height="50">](https://github.com/keturn)<a name="note_11"></a>|@keturn commented|
|-|-|

merged forward to [branches/suite-1638-2,](branches/suite-1638-2,) 
docstrings added.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_12"></a>|@radix set owner to @keturn|
|-|-|

Why is this assigned to me?  It looks like Jerub did a review and forgot to un-mark the "review" tag and reassign it to the author.

**Comment by twisted-trac** [comment]
|[<img alt="keturn's avatar" src="https://avatars.githubusercontent.com/u/83819?s=50" width="50" height="50">](https://github.com/keturn)<a name="note_14"></a>|@keturn set status to `closed`|
|-|-|

(In [17821]) merge suite-1638-2.  Trial discovers test_suite()

Author: jml, keturn
Reviewer: jerub
Fixes [#7250](https://github.com/twisted/twisted/issues/7250)

Adds the ability to trial to override test discovery for a module.  If
need to specially construct a test suite for your module (e.g. to
create data-driven test cases), you can do so in a function named
test_suite().

The name "test_suite" was chosen from its precedence in the Zope3 and
bzr test runners.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
